include/triton/core/tritonserver.h

```diff
@@ -1692,16 +1692,39 @@ TRITONSERVER_DECLSPEC TRITONSERVER_Error*
 TRITONSERVER_ServerOptionsSetCudaMemoryPoolByteSize(
     TRITONSERVER_ServerOptions* options, int gpu_device, uint64_t size);
 
-/// Set the total response cache byte size that the server can allocate in CPU
-/// memory. The response cache will be shared across all inference requests and
-/// across all models.
+// Deprecated. See TRITONSERVER_ServerOptionsSetCacheConfig instead.
+TRITONSERVER_DECLSPEC TRITONSERVER_Error*
+TRITONSERVER_ServerOptionsSetResponseCacheByteSize(
+    TRITONSERVER_ServerOptions* options, uint64_t size);
+
+/// Set the cache config that will be used to initialize the cache
+/// implementation for "cache_name".
+///
+/// Examples:
+///   std::string config_json = R"({"size": 1048576})"
+///   std::string config_json = "{\"size\": 1048576}"
 ///
 /// \param options The server options object.
-/// \param size The total response cache byte size.
+/// FIXME: Included for future extensibility, but not currently used.
+/// \param cache_name The name of the cache.
+/// \param config_json The string representation of config JSON.
 /// \return a TRITONSERVER_Error indicating success or failure.
 TRITONSERVER_DECLSPEC TRITONSERVER_Error*
-TRITONSERVER_ServerOptionsSetResponseCacheByteSize(
-    TRITONSERVER_ServerOptions* options, uint64_t size);
+TRITONSERVER_ServerOptionsSetCacheConfig(
+    TRITONSERVER_ServerOptions* options, const char* cache_name,
+    const char* config_json);
```

Comment:
I am a bit hesitant in adding a new configuration file. I would prefer to have all the control options from either command line or model configuration. This would introduce a third place that the user would need to modify in order to change some of the options.

Reply:
This is a bit tricky, as the cache config is designed to be arbitrary from Triton's POV, it is just passed to the cache implementation to be used however it wants. 

1. For CLI, we would have to either add new fields for every cache and every cache config option we want to officially support, which we don't want to do. Or we can support a generic string CLI argument that supports the same idea of reading arbitrary config from file: `--cache-config "field1=123; field2=456; field3=789"`, but this may be large for some caches or as the scope of config grows. 

I expected to ship the container with default config that should not require too much tweaking for users that want to get going immediately, or is easy to copy-in and replace.

I'm not totally opposed to this option if you think having another file for users to manage is cumbersome.

2. For model config, this is also tricky as we don't currently support unique caches per-model. We would need to map each model to it's own cache with it's own unique config. Maybe this is a route we could go, but I didn't think it was the route we were taking at this time.

Reply:
Thanks for the detailed response! So this is just a byte blob passed from Triton to the cache library? Can you provide some examples for the fields that are currently used for TritonCache and RedisCache? I think one of them is `TRITONSERVER_ServerOptionsSetResponseCacheByteSize`.

Reply:
`TRITONSERVER_ServerOptionsSetResponseCacheByteSize` is going to be deprecated in favor of passing a config that specifies a `size` (or similar) field. ex:
```
{ "response_cache_byte_size": 1048576 }
```
Currently [LocalCache](https://github.com/triton-inference-server/local_cache/pull/1) (the default implementation/shared library) will only expect this size field in a json config (open to change, but it's a convenient format to enforce). New fields may be added in the future.

`RedisCache` will be expecting Redis related configurations such as an IP:Port of a Redis server, some kind of token/cert/authentication, etc. - but those choices can be up to them. ex:
```
{ "redis_server_addr": "localhost:6379", ... }
```
They were also happy with a JSON config file, but functionally it can be any bytes read from a file (when using CLI) or any string (when using the API directly) passed to the `cache_implementation->CacheNew(&cache, cache_config)`. Currently, Triton won't validate or enforce the contents to look any type of way.

Reply:
Unfortunately, we can't deprecate an TRITONSERVER/BACKEND API. The APIs should stay backwards compatible. Would it be possible to define an override mechanism? For example, if the cache byte size is specified in `TRITONSERVER_ServerOptionsSetResponseCacheByteSize` it takes priority over `cache_config`.

>For model config, this is also tricky as we don't currently support unique caches per-model. 

Also, I was wondering who provides the `config_json` file? It seems like currently the cache library would call `TRITONSERVER_ServerOptionsSetCacheConfig`?

Reply:
I can see why you want cache to behave differently for different models, i.e. you want to perform group operation on certain "cache entries" (release the cache if the model is unloaded). Then you can think about adding "tags" to cache entry, which will be a placeholder (no tags defined) for now and extensible in the future.

Reply:
> Then you can think about adding "tags" to cache entry, which will be a placeholder (no tags defined) for now and extensible in the future.

My plan was to eventually add some `TRITONCACHE_CacheEntryAddTags(...)` / `TRITONCACHE_CacheEntryGetTags(...)` for things like flushing specific entries from cache, but didn't include as it wasn't being used currently.

Reply:
> I can see why you want cache to behave differently for different models, i.e. you want to perform group operation on certain "cache entries" (release the cache if the model is unloaded). Then you can think about adding "tags" to cache entry, which will be a placeholder (no tags defined) for now and extensible in the future.

Actually, what I had in mind was a bit different from what you described here. I was thinking about a case where the user wants to use one cache implementation for some models and another implementation for other models (e.g., for large inputs use local cache and for small inputs use redis cache). Right now the user needs to swap the `libtritoncache` in and out of `/opt/tritonserver/caches` to use a different cache library. It means that we cannot ship both of these cache library implementations in the container. I think having the cache config as a part of model configuration would allow us to:

1. Easily change the cache configuration
2. Have a unified place for all the model configuration parameters.

Looks like, in model repository agents we also allow arbitrary key/value paths similar to what Ryan was describing.
https://github.com/triton-inference-server/common/blob/5313d0be99204c89ab696023534043f2ba535868/protobuf/model_config.proto#L1727-L1758

Reply:
@Tabrizian I could see per-model caches as a follow-up refactor (within this release). I agree it's more in-line with the way Backend and RepoAgent APIs are setup and used.

Per-model caches should inherently be less prone to collisions and hopefully involve less evictions per-model being caused by other models (or more if evenly dividing memory but there are unbalanced number of requests coming in per-model, this would be a configuration knob to optimize per use case). This would induce extra thought in how to configure the cache for each model.

Some issues I would want to sort out:

1. If a model is loaded/unloaded, what should we do with the cache? Should this be configurable? Do we sometimes want to keep the cache while a model is unloaded because we plan to load the model again later? Do we sometimes want to delete the cache on model unload because we know we're done with the model? (Possible config options: `keep_loaded: true/false`, `spill_to_disk: true/false`, etc.

2. Do we scrap the concept of server-wide cache metrics in this case and only have per-model cache metrics? 

3. If models each request a certain amount of cache memory, but their total exceeds available memory - is it first come first serve? Should a model fail to load if the cache can not be initialized? (Possible config options: `load_on_cache_fail: true/false`, etc.)

Trying to think of any other potential issues, let me know if you can think of any upfront.

CC @GuanLuo for thoughts on moving to per-model caches + cache_config in model_config

Reply:
Interesting idea, I think it can be a future step if multiple kinds of cache is necessary. I don't think the API needs to be changed to make room for this, everything needed seems to be additions: the API doesn't need to change, just need internal manager to identify and dispatch to the entry points of correct cache library; Add additional way to accept model-specific config (really just pointer to which kind of cache to be used).

Reply:
C API would stay same, but `tritonserver` CLI interaction would have to change. i.e. whether or not to have  `--cache-config`/`--cache-dir` flags, or move to new `config.pbtxt` fields

Reply:
Re Ryan>

1. I think for now we can delete the cache associated with the model on model unloading. For example, the user may upload a new model with different outputs and we may return the old cached results which is not the expected behavior. As you pointed out, we can in the future add new options to the model configuration to further customize cache behavior. 

2. I think we kind of had per-model cache previously too. By including the model name and version in the hash we were expecting to have different caches for each model. For now, I think we can stick to a per-model cache. We can add a "shared cache" concept later by adding more options in the model config if needed.

3. I would say if they have enabled caching but it exceeds the available memory, we should fail the model loading and ask the user to adjust the cache size.

>C API would stay same, but tritonserver CLI interaction would have to change. i.e. whether or not to have --cache-config/--cache-dir flags

I think only the `--cache-config` will be removed but `--cache-dir` would still exist. You can have `/opt/tritonserver/caches/redis`, `/opt/tritonserver/caches/local` and `--cache-dir` would point out to where to find the cache implementations.

Reply:
As discussed offline, we'll keep the concept of a global cache and corresponding cache config, but leave the C APIs designed in a way that allows per-model caches in the future. 

Similarly, will move from JSON config to Protobuf/pbtxt config that can be used standalone for global cache config for now, and then simply be plugged into model `config.pbtxt` section for `response_cache` following the same structure as global cache config to be more intuitive.

Copying response here:

> The CLI will point to the global cache config file, perhaps with a default of `/opt/tritonserver/caches/global_config.pbtxt` or similar. The global cache will not be removed when adding support for per-model caches. The global cache will be used instead if a model enables caching but doesn't specify any per-model cache config. The `--cache-config` CLI flag won't be removed.